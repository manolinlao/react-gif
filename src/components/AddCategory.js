import React, { useState } from 'react'

export const AddCategory = ( { setCategories } ) => {

  const [inputValue,setInputValue] = useState('');

  const handleAdd = () => {
    if(inputValue.trim().length>2){
      setCategories( cats => [...cats,inputValue] );
      setInputValue('');
    }   
  }

  const handleInputChange = ( e ) => {    
    setInputValue( e.target.value );
  }

  const handleSubmit = ( e ) => {
    e.preventDefault();
    if(inputValue.trim().length>2){
      setCategories( cats => [...cats,inputValue] );
      setInputValue('');
    }    
  }
 
  return (
    <>      
      <form onSubmit={ handleSubmit }>
        <input
          type="text"
          value={ inputValue }
          onChange={ handleInputChange }
        />
      </form>
      <button onClick={ handleAdd }>agrega</button>
    </>
  )
}
