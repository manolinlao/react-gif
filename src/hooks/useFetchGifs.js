import { useState, useEffect } from 'react';
import { getGifs } from '../helpers/getGifs';

export const useFetchGifs = ( category ) => {
  console.log('useFetchGifs: category = ' + category);
  const [ state, setState ] = useState(
    {
      data: [],
      loading: true
    }
  );

  useEffect( () => {
    console.log("USEEFFECT EN USEFETCHGIFS");
    getGifs( category )
      .then( imgs => {
        setState(
          {
            data: imgs,
            loading: false
          }
        )
      })
  }, [ category ]);

  console.log('returning state');
  return state; // { data: [], loading: true }
}