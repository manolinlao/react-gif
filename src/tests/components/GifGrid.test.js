import React from 'react';
import { shallow } from 'enzyme';
import { GifGrid } from '../../components/GifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs');

/* 
  Cuando se renderiza por primera vez el componente, la data está vacía
  y el mensaje de loading se ha de mostrar.
  
  Cuando se ha realizado el fetch, el loading ya no se ha de mostrar y 
  tendremos una colección de elementos.
*/

  /*
    Necesitamos mockear la información del resultado del fetch

    Nuestro useFetchçGifs retorna un objeto como este
      {
        data: [],
        loading: true
      }

    Es lo que estaremos mockeando, al  llamar a mockReturnValue lo haremos
    pasándole un objeeto oque es lo que queremos recibir (falseando la respuesta)
  */
describe( 'Pruebas en <GifGrid/>', () => {

  const category = 'One Punch';

  // Hacemos que regrese un array vacío de data y el loading a true
  test( 'debe de mostrarse correctamente', () => {
    
    //mockeo de la respuesta con el estado inicial
    useFetchGifs.mockReturnValue( {
      data: [],
      loading: true
    });

    const wrapper = shallow( <GifGrid category={ category }/>);
    expect( wrapper ).toMatchSnapshot();
  });

  // Prueba recibiendo datos del fetch
  test( 'debe de mostrar items cuando se cargan imágenes de useFetchGifs', () => {
    
    const gifs = [{
      id:'ABC',
      url:'https://localhost/cualquiercosa.jpg',
      title:'cualquier cosa'
    }];

    // mockeo de respuesta con los datos que queremos
    useFetchGifs.mockReturnValue( {
      data: gifs,
      loading: false
    })

    const wrapper = shallow( <GifGrid category={ category }/>);
    expect( wrapper ).toMatchSnapshot();

    // vamos a hacer algunas acepciones más interesantes:
    //    si el loading está en false, hemos de ver que no aparece ya ningún <p>loading</p>
    //    vamos a evaluar que exista <GifGridItem>, tantos como gifs.length tengamos
    expect( wrapper.find('p').exists() ).toBe( false );
    expect( wrapper.find( 'GifGridItem' ).length ).toBe( gifs.length );

  });


});



