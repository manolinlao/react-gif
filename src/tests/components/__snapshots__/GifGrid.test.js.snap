// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifGrid/> debe de mostrar items cuando se cargan im√°genes de useFetchGifs 1`] = `
<div>
  <h3>
    One Punch
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="cualquier cosa"
      url="https://localhost/cualquiercosa.jpg"
    />
  </div>
</div>
`;

exports[`Pruebas en <GifGrid/> debe de mostrarse correctamente 1`] = `
<div>
  <h3>
    One Punch
  </h3>
  <p>
     loading
  </p>
  <div
    className="card-grid"
  />
</div>
`;
